### PR TITLE
Wpe 2.38 quirks

### PR DESCRIPTION
**Pull request by philn**

### CODE REVIEW

**Comment by philn** [comment]
Hold on, I forgot to remove one ifdef...

**Comment by philn** [comment]
> Hold on, I forgot to remove one ifdef...

This fixed, the PR is ready for testing/review!

**Review by emutavchi — COMMENTED** [review]
Quirks should also allow changing playbin flags in `MediaPlayerPrivateGStreamer::setPlaybackFlags` . For example native-audio is Broadcom quirk.

File: Source/WebCore/platform/gstreamer/GStreamerQuirks.h
Comment (emutavchi) [inline Source/WebCore/platform/gstreamer/GStreamerQuirks.h]: Would you mind exposing pipeline configuration via Quirk as well? Currently we maintain a patch to set custom context on the pipeline to signal limited resource usage configuration: https://code.rdkcentral.com/r/plugins/gitiles/rdk/components/generic/rdk-oe/meta-rdk-ext/+/refs/heads/rdk-next/recipes-extended/wpe-webkit/files/2.38/comcast-XIONE-12272-configure-video-resource-usage-w.patch#22, it'd be great to have an option to put it into Westeros quirk.

File: Source/WebCore/platform/gstreamer/GStreamerQuirkWesteros.cpp
Comment (emutavchi) [inline Source/WebCore/platform/gstreamer/GStreamerQuirkWesteros.cpp]: I believe this was removed in favor for change in playbin3 for supporting decoding sinks. For reference: https://github.com/WebPlatformForEmbedded/WPEWebKit/pull/1093. I guess it should be safe to keep this in Westeros quirk until playbin3 changes are sorted out.

File: Source/WebCore/platform/gstreamer/GStreamerQuirkAmLogic.cpp
Comment (emutavchi) [inline Source/WebCore/platform/gstreamer/GStreamerQuirkAmLogic.cpp]: This should check and configure only GstAmlHalAsink. Otherwise it produces a lot of noise:

```
Mar 19 20:33:27 apache-4k HtmlApp-0[1707]: (WPEWebProcess:20): GLib-GObject-WARNING **: 16:33:27.940: g_object_set_is_valid_property: object class 'GstURIDecodeBin' has no property named 'disable-xrun'                                                                                                                   Mar 19 20:33:27 apache-4k HtmlApp-0[1707]: (WPEWebProcess:20): GLib-GObject-WARNING **: 16:33:27.942: g_object_set_is_valid_property: object class 'WebKitWebSrc' has no property named 'disable-xrun'                                                                                                                      
Mar 19 20:33:27 apache-4k HtmlApp-0[1707]: (WPEWebProcess:20): GLib-GObject-WARNING **: 16:33:27.942: g_object_set_is_valid_property: object class 'GstTypeFindElement' has no property named 'disable-xrun'                                                                                
Mar 19 20:33:28 apache-4k HtmlApp-0[1707]: (WPEWebProcess:20): GLib-GObject-WARNING **: 16:33:28.307: g_object_set_is_valid_property: object class 'GstDecodeBin' has no property named 'disable-xrun'
Mar 19 20:33:28 apache-4k HtmlApp-0[1707]: (WPEWebProcess:20): GLib-GObject-WARNING **: 16:33:28.307: g_object_set_is_valid_property: object class 'GstTypeFindElement' has no property named 'disable-xrun'                                                                                                                
Mar 19 20:33:28 apache-4k HtmlApp-0[1707]: (WPEWebProcess:20): GLib-GObject-WARNING **: 16:33:28.307: g_object_set_is_valid_property: object class 'GstQueue2' has no property named 'disable-xrun'                                                                                                                         
Mar 19 20:33:28 apache-4k HtmlApp-0[1707]: (WPEWebProcess:20): GLib-GObject-WARNING **: 16:33:28.334: g_object_set_is_valid_property: object class 'GstQTDemux' has no property named 'disable-xrun'                                                                                                                        
Mar 19 20:33:28 apache-4k HtmlApp-0[1707]: (WPEWebProcess:20): GLib-GObject-WARNING **: 16:33:28.337: g_object_set_is_valid_property: object class 'GstMultiQueue' has no property named 'disable-xrun'                                                                                                                     
Mar 19 20:33:28 apache-4k HtmlApp-0[1707]: (WPEWebProcess:20): GLib-GObject-WARNING **: 16:33:28.339: g_object_set_is_valid_property: object class 'GstH264Parse' has no property named 'disable-xrun'                                                                                                                      
Mar 19 20:33:28 apache-4k HtmlApp-0[1707]: (WPEWebProcess:20): GLib-GObject-WARNING **: 16:33:28.341: g_object_set_is_valid_property: object class 'GstCapsFilter' has no property named 'disable-xrun'                                                                                                                     
Mar 19 20:33:28 apache-4k HtmlApp-0[1707]: (WPEWebProcess:20): GLib-GObject-WARNING **: 16:33:28.344: g_object_set_is_valid_property: object class 'avdec_h264' has no property named 'disable-xrun'                                                                                                                        
Mar 19 20:33:28 apache-4k HtmlApp-0[1707]: (WPEWebProcess:20): GLib-GObject-WARNING **: 16:33:28.349: g_object_set_is_valid_property: object class 'GstAacParse' has no property named 'disable-xrun'                                                                                                                       
Mar 19 20:33:28 apache-4k HtmlApp-0[1707]: (WPEWebProcess:20): GLib-GObject-WARNING **: 16:33:28.374: g_object_set_is_valid_property: object class 'avdec_aac' has no property named 'disable-xrun'                                                                                                                  
Mar 19 20:33:28 apache-4k HtmlApp-0[1707]: (WPEWebProcess:20): GLib-GObject-WARNING **: 16:33:28.401: g_object_set_is_valid_property: object class 'GstInputSelector' has no property named 'disable-xrun'                                                                                                                  
Mar 19 20:33:28 apache-4k HtmlApp-0[1707]: (WPEWebProcess:20): GLib-GObject-WARNING **: 16:33:28.402: g_object_set_is_valid_property: object class 'GstInputSelector' has no property named 'disable-xrun'                                                                                                                  Mar 19 20:33:28 apache-4k HtmlApp-0[1707]: (WPEWebProcess:20): GLib-GObject-WARNING **: 16:33:28.403: g_object_set_is_valid_property: object class 'GstTee' has no property named 'disable-xrun'                                                                                                                            
Mar 19 20:33:28 apache-4k HtmlApp-0[1707]: (WPEWebProcess:20): GLib-GObject-WARNING **: 16:33:28.406: g_object_set_is_valid_property: object class 'GstBin' has no property named 'disable-xrun'   
Mar 19 20:33:28 apache-4k HtmlApp-0[1707]: (WPEWebProcess:20): GLib-GObject-WARNING **: 16:33:28.406: g_object_set_is_valid_property: object class 'GstQueue' has no property named 'disable-xrun'                                                                                                                          
Mar 19 20:33:28 apache-4k HtmlApp-0[1707]: (WPEWebProcess:20): GLib-GObject-WARNING **: 16:33:28.406: g_object_set_is_valid_property: object class 'WebKitVideoSink' has no property named 'disable-xrun'                                                                                                                   
Mar 19 20:33:28 apache-4k HtmlApp-0[1707]: (WPEWebProcess:20): GLib-GObject-WARNING **: 16:33:28.408: g_object_set_is_valid_property: object class 'GstBin' has no property named 'disable-xrun'                                                                                                                            
Mar 19 20:33:28 apache-4k HtmlApp-0[1707]: (WPEWebProcess:20): GLib-GObject-WARNING **: 16:33:28.408: g_object_set_is_valid_property: object class 'GstPlaySinkAudioConvert' has no property named 'disable-xrun'                                                                                                           
Mar 19 20:33:28 apache-4k HtmlApp-0[1707]: (WPEWebProcess:20): GLib-GObject-WARNING **: 16:33:28.409: g_object_set_is_valid_property: object class 'GstAudioResample' has no property named 'disable-xrun'                                                                                                                  
Mar 19 20:33:28 apache-4k HtmlApp-0[1707]: (WPEWebProcess:20): GLib-GObject-WARNING **: 16:33:28.409: g_object_set_is_valid_property: object class 'GstAudioConvert' has no property named 'disable-xrun'                                                                                                                   
Mar 19 20:33:28 apache-4k HtmlApp-0[1707]: (WPEWebProcess:20): GLib-GObject-WARNING **: 16:33:28.409: g_object_set_is_valid_property: object class 'GstIdentity' has no property named 'disable-xrun'                                                                                                                       
Mar 19 20:33:28 apache-4k HtmlApp-0[1707]: (WPEWebProcess:20): GLib-GObject-WARNING **: 16:33:28.409: g_object_set_is_valid_property: object class 'GstQueue' has no property named 'disable-xrun' 
```

File: Source/WebCore/platform/gstreamer/GStreamerQuirkAmLogic.cpp
Comment (emutavchi) [inline Source/WebCore/platform/gstreamer/GStreamerQuirkAmLogic.cpp]: I think it should return false for non GstAmlHalAsink  for other Quirks to kick in

File: Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp
Comment (emutavchi) [inline Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp]: It should return null here even if hole punching quirk did not return anything. This will preserve existing behavior we have, where we rely on playbin2 to find a suitable sink for us.

File: Source/WebCore/platform/gstreamer/GStreamerQuirkWesteros.cpp
Comment (philn) [reply Source/WebCore/platform/gstreamer/GStreamerQuirkWesteros.cpp]: Hmm the current code comes from upstream's ifdefs... Was there some mixup in the upstreaming of #1093 @calvaris ?

File: Source/WebCore/platform/gstreamer/GStreamerQuirks.h
Comment (philn) [reply Source/WebCore/platform/gstreamer/GStreamerQuirks.h]: Yeah I can add this code in the Westeros quirk, no problem I think.

**Comment by philn** [comment]
> Quirks should also allow changing playbin flags in `MediaPlayerPrivateGStreamer::setPlaybackFlags` . For example native-audio is Broadcom quirk.

I wasn't aware of that. Yeah definitely, @calvaris already suggested removing those native-video/etc ifdefs. I can take a look.

File: Source/WebCore/platform/gstreamer/GStreamerQuirkWesteros.cpp
Comment (philn) [reply Source/WebCore/platform/gstreamer/GStreamerQuirkWesteros.cpp]: Hopefully all the comments excepted this one should be addressed in https://github.com/WebKit/WebKit/pull/26183

Comment (calvaris) [reply Source/WebCore/platform/gstreamer/GStreamerQuirkWesteros.cpp]: It was not removed, it won't change upstream and after discussing it with GStreamer upstream developers it is the way to go, so current code is correct.

**Comment by philn** [comment]
I've rebased the branch and backported the changes requested by @emutavchi (see most recent commit).

File: Source/WebCore/platform/gstreamer/GStreamerQuirkWesteros.cpp
Comment (calvaris) [reply Source/WebCore/platform/gstreamer/GStreamerQuirkWesteros.cpp]: Ah, this is for 2.38, I was thinking on other versions. For 2.38, the GStreamer patch drafted in https://gitlab.freedesktop.org/gstreamer/gstreamer/-/merge_requests/4391 is in, so we don't need to tweak the template caps. This would be the last version not needing to tweak the caps, from 2.42 onwards, we need to tweak the template caps as the GStreamer patch won't be present.

Comment (calvaris) [reply Source/WebCore/platform/gstreamer/GStreamerQuirkWesteros.cpp]: Summing up, I think @emutavchi is right and we need to remove this from here.

**Review by calvaris — CHANGES_REQUESTED** [review]

File: Source/WebCore/platform/gstreamer/GStreamerQuirkWesteros.cpp
Comment (philn) [reply Source/WebCore/platform/gstreamer/GStreamerQuirkWesteros.cpp]: Done.

**Review by calvaris — APPROVED** [review]
LGTM. What I don't know is if we should wait until we have a confirmation from @emutavchi .

File: Source/WebCore/platform/gstreamer/GStreamerQuirkBroadcom.cpp
Comment (emutavchi) [inline Source/WebCore/platform/gstreamer/GStreamerQuirkBroadcom.cpp]: This is similar to Amlogic quirk complain in previous review.  Returning 'true' here breaks the loop in GStreamerQuirksManager::configureElement(), blocking other potential quirks. On RDK we'll usually have 2 quirks active: platform specific (amlogic or realtek or broadcom) and westeros. Soc specific quirk should not block Westeros or other potential quirks If an element is not recognized by a quirk.  Also, it could be useful for multiple quirks to be able to configure same element (we don't have such use case today though).

File: Source/WebCore/platform/gstreamer/GStreamerQuirkRealtek.cpp
Comment (emutavchi) [inline Source/WebCore/platform/gstreamer/GStreamerQuirkRealtek.cpp]: Same here.

File: Source/WebCore/platform/gstreamer/GStreamerQuirkWesteros.cpp
Comment (emutavchi) [inline Source/WebCore/platform/gstreamer/GStreamerQuirkWesteros.cpp]: And here.

**Review by emutavchi — CHANGES_REQUESTED** [review]
Few more changes required, that were overlooked last time.

File: Source/WebCore/platform/gstreamer/GStreamerQuirkBroadcom.cpp
Comment (philn) [reply Source/WebCore/platform/gstreamer/GStreamerQuirkBroadcom.cpp]: OK, makes sense I think :) Have you tested the suggested changes on the various platforms?

Comment (philn) [reply Source/WebCore/platform/gstreamer/GStreamerQuirkBroadcom.cpp]: Maybe the `GStreamerQuirksManager::configureElement()` could be re-worked to apply every quirk instead of early loop breaks...

Comment (philn) [reply Source/WebCore/platform/gstreamer/GStreamerQuirkBroadcom.cpp]: Does this make sense? https://github.com/WebKit/WebKit/pull/26456

Comment (emutavchi) [reply Source/WebCore/platform/gstreamer/GStreamerQuirkBroadcom.cpp]: > OK, makes sense I think :) Have you tested the suggested changes on the various platforms?

I did test on Amlogic and Realtek so far. Results looks good. I plan to test Broadcom later this week. 

> Does this make sense? https://github.com/WebKit/WebKit/pull/26456

yes, that works for me.

Comment (emutavchi) [reply Source/WebCore/platform/gstreamer/GStreamerQuirkBroadcom.cpp]: @philn, I just noticed that quirks doesn't cover libwebrtc GStreamerVideoDecoderFactory.cpp. Specifically  the code that skips hw elements in passthrough pipeline: https://github.com/WebPlatformForEmbedded/WPEWebKit/blob/wpe-2.38/Source/WebCore/platform/mediastream/libwebrtc/gstreamer/GStreamerVideoDecoderFactory.cpp#L72

and h264 parser requirement https://github.com/WebPlatformForEmbedded/WPEWebKit/blob/wpe-2.38/Source/WebCore/platform/mediastream/libwebrtc/gstreamer/GStreamerVideoDecoderFactory.cpp#L355

Would you mind adding it to this PR?

Comment (philn) [reply Source/WebCore/platform/gstreamer/GStreamerQuirkBroadcom.cpp]: Hmm I wasn't aware of those, seems like they were not upstreamed. Thanks for the heads-up, I'll check this tomorrow.

Comment (philn) [reply Source/WebCore/platform/gstreamer/GStreamerQuirkBroadcom.cpp]: Can you test this? https://github.com/WebKit/WebKit/pull/26506

Comment (emutavchi) [reply Source/WebCore/platform/gstreamer/GStreamerQuirkBroadcom.cpp]: > Can you test this? https://github.com/WebKit/WebKit/pull/26506

Hi @philn, I did test it on Amlogic, Realtek, Broadcom. I took https://github.com/WebKit/WebKit/pull/26456 and https://github.com/WebKit/WebKit/pull/26506 on top of current PR. Results looks good.

**Comment by modeveci** [comment]
@emutavchi , Once you approved this PR, then I think we are ready to merge. @philn let's wait Eugene's approval and then we can continue.

File: Source/WebCore/platform/gstreamer/GStreamerQuirkBroadcom.cpp
Comment (philn) [reply Source/WebCore/platform/gstreamer/GStreamerQuirkBroadcom.cpp]: I've backported these two here.

**Review by emutavchi — APPROVED** [review]
Looks good to me

**Pull request merged**
